File: Algorithms/Python3.x/8-String_to_Integer_(atoi).py
# Runtime: 32 ms, faster than 99.61% of Python3 online submissions for String to Integer (atoi).
# Memory Usage: 13.1 MB, less than 83.28% of Python3 online submissions for String to Integer (atoi).
class Solution:
    def myAtoi(self, str: str) -> int:
        str = str.strip()
        if len(str) == 0: return 0
        tmp, res, i = '0', 0, 0
        if str[0] in '+-': tmp, i = str[0], 1
        for s in str[i:]:
            if s.isdigit():
                tmp += s
            else:
                break
        if len(tmp) > 1:
            res = int(tmp)
        if res > 2 ** 31 - 1:
            return 2 ** 31 - 1
        elif res < -2 ** 31:
            return -2 ** 31
        else:
            return res

File: Algorithms/C++/8-String_to_Integer_(atoi).cpp
// Runtime: 0 ms, faster than 100.00% of C++ online submissions for String to Integer (atoi).
// Memory Usage: 8.4 MB, less than 71.64% of C++ online submissions for String to Integer (atoi).
class Solution {
public:
    int myAtoi(string str) {
        int ret = 0, 
            sign = 1, 
            i = str.find_first_not_of(' '), 
            base = INT_MAX / 10;
        if (i == -1) // empty string
            return 0;
        if (str[i] == '+' || str[i] == '-') {
            sign = str[i] == '+' ?: -1;
            ++i;
        }
        while (isdigit(str[i])) {
            if (ret > base || (ret == base && str[i] - '0' > 7)) 
                return sign > 0 ? INT_MAX : INT_MIN;
            ret = 10 * ret + (str[i++] - '0');
        }
        return sign * ret;
    }
};
